lasagne/layers/special.py

```diff
@@ -544,6 +547,291 @@ def _meshgrid(height, width):
+    these control points are fixed. The output is interpolated with a
+    bilinear transformation.
+
+    Parameters
+    ----------
+    incoming : a :class:`Layer` instance or a tuple
+        The layer feeding into this layer, or the expected input shape. The
+        output of this layer should be a 4D tensor, with shape
+        ``(batch_size, num_input_channels, input_rows, input_columns)``.
+
+    localization_network : a :class:`Layer` instance
+        The network that calculates the parameters of the thin plate spline
+        transformation as the x and y coordinates of the destination offsets of
+        each control point. The output of the localization network must
+        should be a 2D tensor, with shape
+        ``(batch_size, 2 * num_control_points)``
+
+    downsample_factor : float or iterable of float
+        A float or a 2-element tuple specifying the downsample factor for the
+        output image (in both spatial dimensions). A value of 1 will keep the
+        original size of the input. Values larger than 1 will downsample the
+        input. Values below 1 will upsample the input.
+
+    control_points : integer
+        The number of control points to be used for the thin plate spline
+        transformation. These points will be arranged as a grid along the
+        image, so the value must be a perfect square. Default is 16.
+
+    References
+    ----------
+    .. [1]  Max Jaderberg, Karen Simonyan, Andrew Zisserman,
+            Koray Kavukcuoglu (2015):
+            Spatial Transformer Networks. NIPS 2015,
+            http://papers.nips.cc/paper/5854-spatial-transformer-networks.pdf
+    .. [2]  Fred L. Bookstein (1989):
+            Principal warps: thin-plate splines and the decomposition of
+            deformations. IEEE Transactions on
+            Pattern Analysis and Machine Intelligence.
+            http://doi.org/10.1109/34.24792
+    Examples
+    --------
+    Here, we'll implement an identity transform using a thin plate spline
+    transform. First we'll create the destination control point offsets. To
+    make everything invariant to the shape of the image, the x and y range
+    of the image is normalized to [-1, 1] as in ref [1]_. To replicate an
+    identity transform, we'll simply set the offsets to 0. More complicated
+    transformations can easily be implemented using different x and y offsets
+    (importantly, each control point can have it's own pair of offsets).
+
+    >>> import numpy as np
+    >>> import lasagne
+    >>>
+    >>> # We'll set the bias to be the all offsets of 0
+    >>> num_points = 16
+    >>> b = np.zeros(2*num_points)
+    >>>
+    >>> # Create the network
+    >>> W = lasagne.init.Constant(0.0)
+    >>> l_in = lasagne.layers.InputLayer((None, 3, 28, 28))
+    >>> l_loc = lasagne.layers.DenseLayer(l_in, num_units=2*num_points,
+    ...                                   W=W, b=b, nonlinearity=None)
```

Comment:
Thinking again, there's no need to set `b` to a numpy array, you can just set `b = lasagne.init.Constant(0)`, and that's the default for dense layers. Maybe leave out the bias initialization above, and say:
```
# Create the network
# we'll initialize the weights and biases to zero, so it defaults
# to the identity transform (all control point offsets are zero)
>>> W = b = lasagne.init.Constant(0.0)
```